compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +447,308 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+  rewriter.replaceOp(
+      regionOp, newRegionOp.getResults().take_front(regionOp->getNumResults()));
+
+  // Update terminator.
+  Flow::ReturnOp returnOp =
+      cast<Flow::ReturnOp>(newRegionOp.getBody().front().getTerminator());
+  SmallVector<Value> returnedValues(returnOp.getOperands().begin(),
+                                    returnOp.getOperands().end());
+  returnedValues.push_back(result);
+  returnOp.operandsMutable().assign(returnedValues);
+
+  return newRegionOp;
+}
+
+/// Return `true` if the given op post-dominates the dispatch region.
+static bool isAfterRegion(Operation *op, Flow::DispatchRegionOp regionOp) {
+  Operation *ancestor = regionOp->getBlock()->findAncestorOpInBlock(*op);
+  assert(ancestor && "expected that op and regionOp are in the same block");
+  return regionOp->isBeforeInBlock(ancestor);
+}
+
+// Clone a `target` op that is preceding the given dispatch region op into the
+// dispatch region.
+//
+// All uses of the target inside of the dispatch region are replaced with the
+// results of the cloned op.
+//
+// If `updateUsesOutsideOfRegion` is set, all uses of the target op after the
+// dispatch region, are also updated: The target op's results are returned from
+// the dispatch region an used in those places.
+//
+// Example when `updateUsesOutsideOfRegion` is unset:
+//
+// %0 = "some_op"() : () -> (tensor<?xf32>)
+// %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+//   %1 = "another_op"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+//   flow.return %1 : tensor<?xf32>
+// }
+// %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+//
+// In this example, "some_op" will be cloned into the dispatch region and the
+// OpOperand of "another_op" will be replaced:
+//
+// %0 = "some_op"() : () -> (tensor<?xf32>)
+// %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+//   %0_clone = "some_op"() : () -> (tensor<?xf32>)
+//   %1 = "another_op"(%0_clone) : (tensor<?xf32>) -> (tensor<?xf32>)
+//   flow.return %1 : tensor<?xf32>
+// }
+// %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+static FailureOr<Flow::DispatchRegionOp> clonePrecedingOpIntoDispatchRegion(
+    RewriterBase &rewriter, Operation *target, Flow::DispatchRegionOp regionOp,
+    bool updateUsesOutsideOfRegion) {
+  assert(target->isBeforeInBlock(regionOp) &&
+         "expected that target comes first");
+  Block &body = regionOp.getBody().front();
+
+  // Gather all uses of `target`.
+  SmallVector<OpOperand *> usesInsideOfRegion, usesAfterRegion;
+  bool hasUsesBeforeRegion = false;
+  for (OpOperand &use : target->getUses()) {
+    if (regionOp->isProperAncestor(use.getOwner())) {
+      usesInsideOfRegion.push_back(&use);
+    } else {
+      // Collect only uses that post-dominate the region.
+      if (isAfterRegion(use.getOwner(), regionOp)) {
+        usesAfterRegion.push_back(&use);
+      } else {
+        hasUsesBeforeRegion = true;
+      }
+    }
+  }
+
+  // Clone op into dispatch region.
```

Comment:
Just the sheer number of branching and if/else case handling done here is an indication to me that the ordering of things needs to be reconsidered.

Reply:
I think this op will change/evolve as we experiment.